### PR TITLE
Enable Mesos jemalloc and memory profiling support in DC/OS

### PR DESCRIPTION
**Pull request by lava**
## High-level description

This pull request enables usage of the memory profiling support that was introduced with Mesos 1.6.

It consists of three commits:
 - The first consists of preparatory cleanup work
 - The second changes the Mesos build to use the jemalloc memory allocator by default, to improve general performance of the `mesos-master` and `mesos-agent` binaries.
- The third enables support for capturing memory profiles using a libprocess endpoint.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS_OSS-2137](https://jira.mesosphere.com/browse/DCOS_OSS-2137) Enable Mesos memory profiling in DC/OS images.

## Checklist for all PRs

  - [x] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change:
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: 
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and review process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using GitHub's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

**Comment by rukletsov** [comment]
@mesosphere-mergebot bump-ee

**Comment by lava** [comment]
@mesosphere-mergebot bump-ee

**Comment by lava** [comment]
@mesosphere-mergebot bump-ee

**Comment by lava** [comment]
@mesosphere-mergebot label Ready

File: CHANGES.md [hunk 1]
```diff
@@ -16,6 +16,8 @@ Format of the entries must be.
 
 ### Fixed and improved
 
+* Mesos: Expose memory profiling endpoints. (DCOS_OSS-2137)
```
Comment (bbannier) [inline CHANGES.md]: Append, don't prepend.

File: packages/dcos-integration-test/extra/test_mesos.py [hunk 2]
```diff
@@ -181,6 +182,19 @@ def get_ps_output():
         assert len(get_ps_output().split()) <= 4, 'UCR app has more than 4 processes running in its pid namespace'
 
 
+def test_memory_profiling(dcos_api_session):
+    # Test that we can fetch raw memory profiles
+    master_ip = dcos_api_session.masters[0]
+    profile_id = dcos_api_session.get(
```
Comment (bbannier) [inline packages/dcos-integration-test/extra/test_mesos.py]: Do we authorize this or the `GET` below? I didn't see any updates on IAM resource ids. If not, this sounds pretty bad since we could leak sensitive data in `strict` mode.

If we don't authorize, I'd suggest to either add authorization before enabling in Mesos, or only adding the config options in `disabled` security mode.

@jgehrcke

File: packages/dcos-integration-test/extra/test_mesos.py [hunk 3]
```diff
@@ -181,6 +182,19 @@ def get_ps_output():
...
+        '/memory-profiler/start?duration=1secs', host=master_ip, port=5050).json()['id']
+
+    time.sleep(2)
```
Comment (bbannier) [inline packages/dcos-integration-test/extra/test_mesos.py]: Why do we `sleep`. I'd prefer if we wouldn't sleep at all. If you still want to sleep, please explain what drives this particular timeout so it can be understood should this ever become flaky.

File: CHANGES.md [hunk 4]
```diff
@@ -64,6 +64,8 @@ Format of the entries must be.
 
 ### Notable changes
 
+* Mesos now uses the jemalloc memory profiler by default. (DCOS_OSS-2137)
```
[the remark this answers is not in the document]
Comment (lava) [reply CHANGES.md]: :-1: 

The policy is written down at the top of this very file:

> * Add to the top of the existing list.

File: packages/mesos/build [hunk 5]
```diff
@@ -21,6 +21,7 @@ LDFLAGS="-Wl,-rpath,/opt/mesosphere/lib" "/pkg/src/mesos/configure" \
   --enable-libevent --enable-ssl \
   --enable-install-module-dependencies \
   --enable-grpc \
+  --enable-jemalloc-allocator \
```
[the remark this answers is not in the document]
Comment (lava) [reply packages/mesos/build]: I thought about it, but decided against it on the grounds that it would be a massiva PITA with no real benefit until we have a second application also using jemalloc as memory allocator.

File: CHANGES.md (hunk 1 above)
Comment (lava) [reply CHANGES.md]: See above.

File: packages/dcos-integration-test/extra/test_mesos.py (hunk 3 above)
Comment (rukletsov) [reply packages/dcos-integration-test/extra/test_mesos.py]: We sleep because we wait until jemalloc collects some data. I would however explicitly call `/stop` to avoid any flakiness.

Comment (bbannier) [reply packages/dcos-integration-test/extra/test_mesos.py]: If you care about actual data being returned, maybe add some test that the response is not empty or some other metric of usefulness.

File: gen/dcos-config.yaml [hunk 6]
```diff
@@ -456,63 +456,63 @@ package:
...
-      MESOS_LAUNCHER_DIR=/opt/mesosphere/active/mesos/libexec/mesos
-      MESOS_EXECUTOR_ENVIRONMENT_VARIABLES=file:///opt/mesosphere/etc/mesos-executor-environment.json
-      MESOS_EXECUTOR_REGISTRATION_TIMEOUT=10mins
-      MESOS_EXECUTOR_REREGISTRATION_TIMEOUT=10secs
-      MESOS_RECONFIGURATION_POLICY=additive
-      MESOS_RECOVERY_TIMEOUT={{ mesos_recovery_timeout }}
+      GLOG_drop_log_memory=false
       MESOS_CGROUPS_ENABLE_CFS=true
       MESOS_CGROUPS_LIMIT_SWAP=false
+      MESOS_CONTAINER_LOGGER=com_mesosphere_mesos_JournaldLogger
```
[the remark this answers is not in the document]
Comment (lava) [reply gen/dcos-config.yaml]: Fixed.

File: gen/dcos-config.yaml [hunk 7]
```diff
@@ -456,63 +456,63 @@ package:
...
+      MESOS_CONTAINERIZERS=docker,mesos
+      MESOS_DEFAULT_CONTAINER_DNS=file:///opt/mesosphere/etc/mesos-slave-dns.json
       MESOS_DISALLOW_SHARING_AGENT_PID_NAMESPACE=true
+      MESOS_DOCKER_VOLUME_CHECKPOINT_DIR=/var/lib/mesos/isolators/docker/volume
```
[the remark this answers is not in the document]
Comment (lava) [reply gen/dcos-config.yaml]: Fixed.

File: packages/dcos-integration-test/extra/test_mesos.py (hunk 3 above)
Comment (lava) [reply packages/dcos-integration-test/extra/test_mesos.py]: Fixed.

File: packages/mesos/build (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbannier) [reply packages/mesos/build]: Please call this out in the commit message. Somebody might come along in the future wanting to use jemalloc and wonder if there is any reason Mesos shouldn't use an unbundled one.

File: packages/dcos-integration-test/extra/test_mesos.py [hunk 8]
```diff
@@ -181,6 +181,22 @@ def get_ps_output():
         assert len(get_ps_output().split()) <= 4, 'UCR app has more than 4 processes running in its pid namespace'
 
 
+def test_memory_profiling(dcos_api_session):
+    # Test that we can fetch raw memory profiles
+    master_ip = dcos_api_session.masters[0]
+    r0 = dcos_api_session.get(
+        '/memory-profiler/start', host=master_ip, port=5050)
+    assert r0.status_code == 200
```
Comment (bbannier) [inline packages/dcos-integration-test/extra/test_mesos.py]: We should print the body on assertion failures, here and below.

    assert r0.status_code == 200, r0.text

File: packages/dcos-integration-test/extra/test_mesos.py [hunk 9]
```diff
@@ -181,6 +181,22 @@ def get_ps_output():
...
+
+    r1 = dcos_api_session.get(
+        '/memory-profiler/stop', host=master_ip, port=5050)
+    assert r1.status_code == 200
+
+    r2 = dcos_api_session.get(
+        '/memory-profiler/download/raw', host=master_ip, port=5050)
+    assert r2.status_code == 200
```
Comment (bbannier) [inline packages/dcos-integration-test/extra/test_mesos.py]: We could check that `r2.text` is not empty or some other indication that the response was as intended.

Comment (lava) [reply packages/dcos-integration-test/extra/test_mesos.py]: Let's not assume any specific format for the heap profiles. Since i'm not leaving any time for measurements in order to avoid sleeping, an empty response doesn't necessarily indicate failure.

File: packages/dcos-integration-test/extra/test_mesos.py (hunk 8 above)
Comment (lava) [reply packages/dcos-integration-test/extra/test_mesos.py]: What do you expect to see in the body? "File Not Found"? Any interesting output would be in the web server logs.

Comment (bbannier) [reply packages/dcos-integration-test/extra/test_mesos.py]: A request could file at a number of places as it finds its way to the master. Some of these response return additional data on top of the status code. Outputting that (sometimes redundant) text here can help genug issues faster.

We also do that in a couple of places in this test already, uniformity yadadada.

Comment (lava) [reply packages/dcos-integration-test/extra/test_mesos.py]: Ok, done.

File: packages/mesos/build (hunk 5 above)
[the remark this answers is not in the document]
Comment (lava) [reply packages/mesos/build]: Done.

**Comment by lava** [comment]
@mesosphere-mergebot bump-ee

**Comment by rukletsov** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/azure/hybrid-cloud https://jira.mesosphere.com/browse/COPS-3583

File: packages/dcos-integration-test/extra/test_mesos.py [hunk 10]
```diff
@@ -181,6 +181,22 @@ def get_ps_output():
         assert len(get_ps_output().split()) <= 4, 'UCR app has more than 4 processes running in its pid namespace'
 
 
+def test_memory_profiling(dcos_api_session):
+    # Test that we can fetch raw memory profiles
```
Comment (bbannier) [inline packages/dcos-integration-test/extra/test_mesos.py]: We could have used a Python docstring here and terminated the comment with a period.

**Comment by lava** [comment]
@mesosphere-mergebot bump-ee

**Comment by lava** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/dcos-e2e/docker/static/strict https://jira.mesosphere.com/browse/DCOS_OSS-2115

**Comment by rukletsov** [comment]
@mesosphere-mergebot override-status teamcity/dcos/test/dcos-docker/static https://jira.mesosphere.com/browse/DCOS_OSS-2115

**Comment by rukletsov** [comment]
@mesosphere-mergebot label Ship It
